### PR TITLE
lib/log: write test output to testing.TB

### PR DESCRIPTION
**Pull request by bobheadxi**
This writes test output to t.Log, which allows us to integrate with 'go test' output management. I think this closes https://github.com/sourcegraph/sourcegraph/pull/35299

Also introduces `logtest.ScopedWith(t testing.TB, options LoggerOptions)` with additional configuration.

Related discussion: https://sourcegraph.slack.com/archives/C3B3SDBMY/p1652432949068519

**edit** - Actually I realized I implemented the above wrong. When I fixed it, I realized -v already works hiding output out-of-the-box (notice similar behaviour with -v in screenshots below). Do we need to do this? => https://github.com/sourcegraph/sourcegraph/pull/35430#issuecomment-1127291504

## Test plan

Before:

image

After:

image

~old screenshot: https://user-images.githubusercontent.com/23356519/168342703-005445b4-16d4-45c5-8679-16097d8ef5ae.png~

### CODE REVIEW

**Comment by keegancsmith** [comment]
> Actually I realized I implemented the above wrong. When I fixed it, I realized -v already works hiding output out-of-the-box (notice similar behaviour with -v in screenshots below). Do we need to do this?

go has different behaviour when running just plain old `go test` with no arguments. ie, just test the pkg of $PWD. Then it will include the stderr output by default. For my editor (and I assume others) this is the default behaviour of running go tests with a shortcut, just test the current file's pkg.

**Comment by bobheadxi** [comment]
Acked - I've more carefully re-implemented the test writer in https://github.com/sourcegraph/sourcegraph/pull/35430/commits/375b00c0b5b0739fd76a63f4eef4437999c08c03 and now we have formatting consistent with dev output that goes to `t.Logf`:

image

The only caveat is we now get the wrapping implementation's callsite (`core.go:73`) in this case on all log output, with no way of configuring that it seems. Is that okay?

File: lib/log/logtest/logtest.go [hunk 1]
```diff
@@ -63,27 +63,57 @@ type CapturedLog struct {
...
-// Get retrieves a logger from scoped to the the given test.
-//
-// Unlike log.Scoped(), logtest.Scoped() is safe to use without initialization.
-func Scoped(t testing.TB) log.Logger {
+type LoggerOptions struct {
+	// Level configures the minimum log level to output.
+	Level log.Level
+	// FailOnErrorLogs indicates that the test should fail if an error log is output.
+	FailOnErrorLogs bool
+}
```
Comment (keegancsmith) [inline lib/log/logtest/logtest.go:71]: what is an example package you would expect to use these logger options? I'm surprised you need to expand the API, it looks unrelated to the stated purpose of this PR.

File: lib/log/internal/encoders/config.go [hunk 2]
```diff
@@ -62,3 +62,20 @@ func ApplyDevConfig(cfg zapcore.EncoderConfig) zapcore.EncoderConfig {
...
+	config := OpenTelemetryConfig
+	if development {
+		config = applyDevConfig(config)
+	}
+
+	switch format {
+	case OutputConsole:
+		enc = zapcore.NewConsoleEncoder(config)
+	case OutputJSON:
+		enc = zapcore.NewJSONEncoder(config)
```
Comment (keegancsmith) [inline lib/log/internal/encoders/config.go:76]: nit: rather than mutating enc and having a named variable, just return here.

**Review by keegancsmith — APPROVED** [review]
LGTM, nice. The core.go always appearing is an acceptable trade off I'd say. Also took a quick look to see if we could get rid of it, we can't. testing.TB.Helper doesn't help.

File: lib/log/logtest/core.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
```
Comment (unknwon) [inline lib/log/logtest/core.go:46]: nit: I think these are for the file header not the function header, but honestly? I have no idea. 😂

**Review by unknwon — APPROVED** [review]

File: lib/log/logtest/logtest.go (hunk 1 above)
Comment (bobheadxi) [reply lib/log/logtest/logtest.go:71]: A `logtest.Scoped` that enables level configuration seems useful, since you pointed out there are many cases of `log15.DiscardHandler` that can now use this

File: lib/log/internal/encoders/config.go (hunk 2 above)
Comment (bobheadxi) [reply lib/log/internal/encoders/config.go]: Ah thanks! 7d50bb12a090c52fa0b1a59301d9d4265f927a6e

File: lib/log/logtest/core.go (hunk 3 above)
Comment (bobheadxi) [reply lib/log/logtest/core.go:46]: That's true, but I only copied a small part so oh well 😛 It does only say:

```
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
```

So I guess it can go anywhere, and I felt this was a bit more readable and less confusing 😆

**Review by jhchabran — APPROVED** [review]
